### PR TITLE
dialects: (bufferization) Add `to_memref` op

### PR DESCRIPTION
**Pull request by n-io**

### CODE REVIEW

File: tests/filecheck/mlir-conversion/with-mlir/dialects/bufferization/ops.mlir [hunk 1]
```diff
@@ -3,11 +3,13 @@
 module{
     %t = "bufferization.alloc_tensor"() <{"operandSegmentSizes" = array<i32: 0, 0, 0>}> : () -> tensor<10x20x30xf64>
     %m = "test.op"() : () -> memref<30x20x10xf32>
-    %m_t = "bufferization.to_tensor"(%m) : (memref<30x20x10xf32>) -> tensor<30x20x10xf32>
+    %m_t = "bufferization.to_tensor"(%m) <{"restrict"}> : (memref<30x20x10xf32>) -> tensor<30x20x10xf32>
```
Comment (superlopuh) [inline tests/filecheck/mlir-conversion/with-mlir/dialects/bufferization/ops.mlir:6]: might as well write this in custom syntax also IMO

File: xdsl/dialects/bufferization.py [hunk 2]
```diff
@@ -13,15 +13,46 @@
...
+    def infer(self, constraint_context: ConstraintContext) -> Attribute:
+        if self.name in constraint_context.variables:
+            return constraint_context.variables[self.name]
+        if self.name == "T" and "M" in constraint_context.variables:
+            m_type = constraint_context.variables["M"]
+            if isa(m_type, MemRefType[Attribute]):
+                return TensorType(m_type.get_element_type(), m_type.get_shape())
+            if isa(m_type, UnrankedMemrefType[Attribute]):
+                return UnrankedTensorType(m_type.element_type)
+        raise ValueError(f"Unexpected {self.name} - cannot infer attribute")
```
Comment (superlopuh) [inline xdsl/dialects/bufferization.py:55]: Very nice

File: tests/filecheck/mlir-conversion/with-mlir/dialects/bufferization/ops.mlir (hunk 1 above)
Comment (n-io) [reply tests/filecheck/mlir-conversion/with-mlir/dialects/bufferization/ops.mlir]: Done as suggested

File: tests/filecheck/mlir-conversion/with-mlir/dialects/bufferization/ops.mlir [hunk 3]
```diff
@@ -3,11 +3,13 @@
 module{
     %t = "bufferization.alloc_tensor"() <{"operandSegmentSizes" = array<i32: 0, 0, 0>}> : () -> tensor<10x20x30xf64>
     %m = "test.op"() : () -> memref<30x20x10xf32>
-    %m_t = "bufferization.to_tensor"(%m) : (memref<30x20x10xf32>) -> tensor<30x20x10xf32>
+    %m_t = bufferization.to_tensor"(%m) restrict writable : (memref<30x20x10xf32>) -> tensor<30x20x10xf32>
```
Comment (superlopuh) [inline tests/filecheck/mlir-conversion/with-mlir/dialects/bufferization/ops.mlir:6]:
```suggestion
    %m_t = bufferization.to_tensor (%m) restrict writable : (memref<30x20x10xf32>) -> tensor<30x20x10xf32>
```

**Comment by n-io** [comment]
reworked as suggested without depending on constraint variable names

File: xdsl/dialects/bufferization.py [hunk 4]
```diff
@@ -14,14 +14,61 @@
...
     result_def,
     var_operand_def,
 )
+from xdsl.utils.exceptions import VerifyException
+from xdsl.utils.hints import isa
+
+
+class TensorMemrefInferenceConstraint(VarConstraint[Attribute]):
+    """
+    Constraint to infer tensor shapes from memref shapes.
```
Comment (superlopuh) [inline xdsl/dialects/bufferization.py:33]: Could you please add a description of how the mechanism below works for inference and verification?

**Review by superlopuh — APPROVED** [review]

**Pull request merged**